client/src/sge/GraphicsGeometry.cpp

```diff
@@ -329,6 +443,125 @@ namespace sge {
         return textureIdx[textureAbsolutePath];
     }
 
+    /**
+     * Load an animation from an assimp scene object and add it to
+     * ModelComposite data structures
+     * @param animation
+     * @return
+     */
+    ModelComposite::Animation ModelComposite::loadAnimation(const aiAnimation &animation) {
+        Animation cur;
+        cur.duration = animation.mDuration;
+        cur.ticksPerSecond = animation.mTicksPerSecond;
+        for (unsigned int i = 0; i < animation.mNumChannels; i++) {
+            std::string nodeName = animation.mChannels[i]->mNodeName.C_Str();
+            // Load missing bones in animation
+            if (!boneMap.count(nodeName)) {
+                boneMap[nodeName] = boneMap.size();
+                bones.offsetMatrices.push_back(glm::mat4(1)); // To maintain invariant that this vector can always be indexed by bone id, tho this matrix (should) never be used
+                numBones++;
+            }
+            int id = boneMap[nodeName];
+            cur.channels[id] = BonePose(*animation.mChannels[i], id);
+        }
+        return cur;
+    }
+
+    /**
+     * PRECONDITION: All meshes already loaded (and by extension, all bones directly attached to vertices have been loaded)
+     *
+     * Loads the model's bone hierarchy and load in any missing bones
+     * @param root aiNode to build from
+     * @return Resulting tree translated to ModelComposite::BoneNode's
+     */
+    ModelComposite::BoneNode ModelComposite::buildBoneHierarchy(aiNode *root) {
+        BoneNode cur;
+        std::string boneName = root->mName.C_Str();
+        // Assimp is dumb and adds a bunch of extra nodes in the hierarchy tree with identity matrix relative transformations :/
+        int boneId = -1;
+        if (boneMap.count(boneName)) {
+            boneId = boneMap[boneName];
+        }
+        cur.relativeTransform = assimpToGlmMat4(root->mTransformation);
+        cur.id = boneId;
+        for (unsigned int i = 0; i < root->mNumChildren; i++) {
+            cur.children.push_back(buildBoneHierarchy(root->mChildren[i]));
+            // Delete unnecessary nodes
+            if (cur.children.back().id == -1 && cur.children.back().children.empty()) {
+                cur.children.pop_back();
+            }
+        }
+
+        return cur;
+    }
+
+    /**
+     * Return model pose for a given animation at a given timestamp
+     * @param animationId Model's animation identifier
+     * @param time Timestamp to retrieve pose
+     * @return Model's pose at given timestamp
+     */
+    void ModelComposite::animationPose(int animationId, float time, ModelPose& outputModelPose) {
+        assert(animationId >= -1  && animationId < animations.size() && animated);
```

Comment:
For some reason, the check animationId < animations.size() is failing for me, even though I have 
animationId = -1
animations.size() = 1

Reply:
ye idk wtf is going on here

Reply:
Hmm maybe animations.size() is returning an unsigned int, and animationId is a signed int, and c++ is just casting very strangely?? I will look into this.

Reply:
Likely that, if i remember correctly signed ints are converted to unsigned int's in these types of conversions

Reply:
Fixed. Well that's a good reminder for me to think carefully about types, and also run in debug mode occasionally even though it's super slow